### ISSUE TITLE
Uncaught TypeError: jQuery(...).ZeroSpamDavidWalsh is not a function

### ISSUE THREAD

**Issue by ianpegg** [Highfivery/zero-spam-for-wordpress#345]
**Describe the bug**
I've noticed an issue that only occurs when logged out. At first, I thought it might be a caching issue so I tested in my development environment without any form of caching enabled and the problem persisted.

This error pops up in the console when the David Walsh Technique is _disabled_. The davidwalsh.js script is therefore not loaded, however the following inline script is still appended to the footer:

`jQuery(".wpcf7-form").ZeroSpamDavidWalsh()`

Naturally, when this is executed the JS attempts to call a function that doesn't exist. If the DW Technique is _enabled_ then this error goes away. Clearly, the plugin is not enqueuing the DW script when it isn't needed - it also needs to ensure that the accompanying inline script is also not enqueued.

This issue occurs consistently as described above across all sites I manage that use the latest version of Zero Spam. All other plugins and WP core are currently up to date. The server is running PHP 8.0.

**To Reproduce**
1. Go to 'Settings' --> 'Zero Spam'
2. Click on 'David Walsh' in the side menu
3. Click the toggle to enable the DW Technique (note: the toggle label says 'Enabled' whether it is actually enabled or not)
4. Hit 'Save Settings'. No need to add a custom CSS class to the field if you are using CF7, as I am.

**Expected behavior**
When the DW Technique is disabled, no JS should be enqueued or added inline to the front-end of the site, whether logged in or not.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: macOS Big Sur v11.7
 - Browser: Chrome
 - Version: 107.0.5304.121 (Official Build) (x86_64) 

**Smartphone (please complete the following information):**

**Additional context**

**Comment by bmarshall511** [comment]
Duplicate of https://github.com/Highfivery/zero-spam-for-wordpress/issues/343

**Comment by ianpegg** [comment]
Hi @bmarshall511 - this isn't a duplicate of #343. I checked that report before posting mine.

In #343, @leewillis77 reported an issue when jQuery is deferred or footer loaded but I am not having that issue as I don't do that with my sites. The OP is also _using_ the DW technique, which I am not. I'm also _not_ having a problem with form submissions - they work just fine for me.

Thanks for your latest patches. Unfortunately, my original issue is still occurring after applying the latest update.

Thanks for your help!

**Comment by bmarshall511** [comment]
@ianpegg Dug into this a little more, but unable to replicate it using the default implementation. Can you provide the URL of where you're seeing this JS error?

**Comment by bmarshall511** [comment]
@ianpegg It seems that the issue you're experiencing is specific to your setup. Based on the information provided, it appears that your JavaScript files are being concatenated, causing the problem. In order for us to investigate further, we would need to replicate the issue using your specific site configuration.

If you could provide more details about your optimization techniques, such as JS concatenation and any other relevant information about your setup, we can try to recreate the issue and provide a solution. However, if you have a Zero Spam license key, it would be best to [reach out to our support team](https://www.zerospam.org/contact/). One of our engineers will be able to conduct a more thorough investigation and assist you in resolving the issue.

In the meantime, we recommend either disabling the concatenation for the Zero Spam scripts or reorganizing the concatenation result to ensure that the scripts are concatenated in the correct order.

**Comment by bmarshall511** [comment]
@ianpegg Haven't received anything yet from the Zero Spam contact form. Can you try to resend it? If possible, please include a login so we can dig into what you have going on in the backend. Otherwise, there's not much we'll be able to tell with just a front-facing page.

If you're seeing `jQuery(".wpcf7-form").ZeroSpamDavidWalsh();` then the `zerospam-davidwalsh` should be enqueued. If you're using a third-party plugin that is reorganizing the scripts in any kind of way, this could be your issue. Have you disabled all optimization plugins & anything that would effect the default loading of scripts and tested? We've tried numerous times on existing installs and fresh installs to replicate with no luck, so likely an issue with your particular setup.

**Comment by ianpegg** [comment]
Hello @bmarshall511 

I've now found some time to test this in my development environment. I deactivated all plugins and mu-plugins and switched to the 2021 theme. It was only when I re-enabled CF7 that the issue reoccurred, which lead me down the path of reviewing the code in the [class-contactform7.php module](https://github.com/Highfivery/zero-spam-for-wordpress/blob/master/modules/contactform7/class-contactform7.php).

I noticed that the ['add_scripts' function in this module](https://github.com/Highfivery/zero-spam-for-wordpress/blob/master/modules/contactform7/class-contactform7.php#L55) checks whether 'verify_contactform7' is enabled before then enqueuing the 'zerospam-davidwalsh' script and also injecting into the footer the inline script which has been causing such consternation for all this time.

The problem is that this logic assumes that if the user has enabled "Protect CF7 Submissions" in the settings then this also means they want the David Walsh technique enabled too. This is demonstrated by the fact that this function doesn't check to see whether 'davidwalsh' is also enabled and so then proceeds to enqueue the script and inject the footer script. Whilst the 'zerospam-davidwalsh' script can only be successfully enqueued if it is first registered by the module inside 'class-davidwalsh.php', the inline script is added to the WP footer regardless, causing the JS error at runtime as described in the original post.

To clarify, the [DavidWalsh module correctly checks](https://github.com/Highfivery/zero-spam-for-wordpress/blob/master/modules/davidwalsh/class-davidwalsh.php#L35) to see whether the module has been activated in the settings before attempting to register the 'zerospam-davidwalsh'. Therefore, it will not register the script if the module is deactivated, meaning that when the CF7 module comes to enqueue the script, nothing is enqueued.

I modified the CF7 module to check that 'davidwalsh' is enabled by adding the following check to the conditional in the '[add_scripts](https://github.com/Highfivery/zero-spam-for-wordpress/blob/master/modules/contactform7/class-contactform7.php#L55)' function:

`'enabled' === \ZeroSpam\Core\Settings::get_settings( 'davidwalsh' )`

With this line added, the plugin now respects the user's choice if they choose to activate "Protect CF7 Submissions" but deactivate the "David Walsh Technique".

Hope this helps!
